### DISCUSSION TITLE
Search Analytics

### DISCUSSION THREAD

**Discussion by Fringie** [microsoft-search/pnp-modern-search#2772] [Ideas]
Hi guys,

I'm building some analytics functionality for PNP Modern Search; whenever a user searches for something, we will track what the user searched for, what the user selects and if the user couldn't find what they were looking for. I've got a few ideas on implementing this, and there is potential for this to be contributed to PNP Modern Search (I will discuss it with my boss tomorrow).

In terms of implementation, what I'm currently thinking is:

1. Add logic to the search box submit an event; we track what the user searches for and store the data in the current session
2. Add a mutation observer to the page; this will track the search results the user clicks - store this in the current session
3. On page unload, save the entry to a SharePoint list (administrators use PowerBI to traverse the data)
3.1 We might use Dataverse to store the data, if we track analytics at site collection level, we will use SharePoint lists, if we decide to track analytics at the tenant level, we will use dataverse. This is part of a bigger piece of work so it may be that we start with SharePoint lists and then add a dataverse option in, so you can do it at the site collection level or dataverse.
3.2 If the browser unexpectedly crashes etc, then the analytics won't be saved. I may make the logic save the data after each click

Any thoughts? I'm curious to see what you guys would be looking for or would expect.

**Comment by wobba** [comment]
Good idea and as long as you anonymize what you log you should be ok. I would probably add logging directly to some datastore directly via async calls on the clicks - which is how Microsoft does it in their OOB offerings.

You could perhaps use the API for PowerBI and it directly instead of some middle storage? Using SP for storage seems like a hack to me.

**Reply by Fringie** [reply]
> Good idea and as long as you anonymize what you log you should be ok.

Good point, we are tracking the individual users, so we can understand in full detail, why users are not able to find documents, whether it's a training problem, if the search experience needs to be changed and so on. There's a GDPR consideration, so I will need to delete logs in a similar fashion to how the M365 audit log works I imagine.


> You could perhaps use the API for PowerBI and it directly instead of some middle storage? 

Can you clarify, I'm not aware of how this would work. We need some form of persistent storage, and I'm not aware of PowerBI having that capability.

>I would probably add logging directly to some datastore directly via async calls on the clicks

Yeah, in hindsight that is a better approach.

>Using SP for storage seems like a hack to me.

I wouldn't say it's a hack, just an.. outdated approach :D (SharePoint Classic would store audit logs on the site collection, but now it's all in the compliance admin centre). But fair point, SharePoint isn't the place for this data, we'll store it in Dataverse.


I spoke with my boss, and they said it should be fine to make this feature open-source.

**Reply by wobba** [reply]
You can look at push datasets with PowerBI (https://www.sqlbi.com/articles/creating-real-time-dashboards-in-power-bi-with-push-datasets/). Haven't looked at them myself though :)

**Comment by JeighBI** [comment]
@Fringie - were you able to make any progress on this? I'm interested in collecting search audit data when using the PnP web parts but I'm not sure if there is anything available.
@wobba - is there a way to get search reports like "Top queries" to work for these web parts?
Thanks!

**Reply by wobba** [reply]
No, as there is no API to push back clicks or record queries.
